fendermint/vm/message/src/ipc.rs

```diff
@@ -0,0 +1,169 @@
+#[allow(clippy::large_enum_variant)]
+pub enum IpcMessage {
+    /// A bottom-up checkpoint coming from a child subnet "for resolution", relayed by a user of the parent subnet for a reward.
+    ///
+    /// The reward can be given immediately upon the validation of the quorum certificate in the checkpoint,
+    /// or later during execution, once data availability has been confirmed.
+    BottomUpResolve(SignedRelayedMessage<CertifiedMessage<BottomUpCheckpoint>>),
+
+    /// A bottom-up checkpoint proposed "for execution" by the parent subnet validators, provided that the majority of them
+    /// have the data available to them, already resolved.
+    ///
+    /// To prove that the data is available, we can either use the ABCI++ "process proposal" mechanism,
+    /// or we can gossip votes using the _IPLD Resolver_ and attach them as a quorum certificate.
+    BottomUpExec(CertifiedMessage<BottomUpCheckpoint>),
+
+    // TODO
+    TopDown,
+}
+
+/// A message relayed by a user on the current subnet.
+///
+/// The relayer pays for the inclusion of the message in the ledger,
+/// but not necessarily for the execution of its contents.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct RelayedMessage<T> {
+    /// The relayed message.
+    pub message: T,
+    /// The address (public key) of the relayer in the current subnet.
+    pub relayer: Address,
+    /// The nonce of the relayer in the current subnet.
+    pub sequence: u64,
+}
+
+/// Relayed messages are signed by the relayer, so we can rightfully charge them message inclusion costs.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct SignedRelayedMessage<T> {
+    /// The relayed message with the relayer identity.
+    pub message: RelayedMessage<T>,
+    /// The signature of the relayer, for cost and reward attribution.
+    pub signature: Signature,
+}
+
+/// A message with a quorum certificate from a group of validators.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct CertifiedMessage<T> {
+    /// The message the validators signed.
+    pub message: T,
+    /// The quorum certificate.
+    pub certificate: MultiSig,
+}
+
+/// A quorum certificate consisting of a simple multi-sig.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct MultiSig {
+    pub signatures: Vec<ValidatorSignature>,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct ValidatorSignature {
+    pub validator: Address,
+    pub signature: Signature,
+}
+
+/// A periodic bottom-up checkpoints contains the source subnet ID (to protect against replay attacks),
+/// a block height (for sequencing), any potential handover to the next validator set, and a pointer
+/// to the messages that need to be resolved and executed by the parent validators.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct BottomUpCheckpoint {
```

Comment:
We need a few more fields as part of the `BottomUpCheckpoint`: https://github.com/consensus-shipyard/ipc-solidity-actors/blob/cc109a57d62755cf849028a366258c38c04e49c8/src/structs/Checkpoint.sol#L8. Mainly:
- `fee`: Fees to be distributed in the parent (this may change for Fendermint).
- `prevHash`: The hash of the previous checkpoint to build a chain of checkpoints that can be persisted and discoverable externally (IPFS/FIL sector).
- `proof` --> `snapshot`: We should rename it to snapshot, but whatever state snapshot or proof of the state a subnet wants to propagate to its parent.

Reply:
Thank you for commenting on this, it is something I needed your input on. 

I see that what you are pointing at is in Solidity, so I was wondering if we can put that behind the CID, and keep only the fields here that we need in order to validate in the SubnetActor that this checkpoint meta-data is correct, and it can be resolved into the full thing. 

For example the fields you listed:
* `fee`: I assume each transaction pays its own fee in our chosen default execution model, where everyone pays for their own stuff. Is this fee a general fee to pay for the checkpoint itself? See if we can leave it behind the CID. I thought that we can "execute" this meta-data message by sending it to the SubnetActor for i) validation and ii) taking payment from the relayer; if fees has to be sent from the common pot to some parent validators, indeed this would be a good time to do it, and for that it should be visible in this record. Could it not live in the SubnetActor instead, so that it automatically does it, instead of the subnet validators sending a different value each time?
* `prevHash`: Do we really need this to be in the meta-data for validation? We have a height and if it's signed by the right validator set at the right height, should we not accept that this is what the committee is saying the next checkpoint is? I'm not saying we don't need it, but can it be behind the CID?
* `snapshot`: This sounds like it can be behind the CID, that it's just like cross-messages, but different. One ramification of lifting it into the meta-data is that I built the resolver in #201 by assuming a single CID to be resolved, not a list - I assumed we can always create a struct to hold multiple CIDs, which makes it easier to track its completion. We can still know about it, and treat it in a special way during the execution of the checkpoint, when this is already available, otherwise we would just stuff it in the blockstore without even charging for it. 

I am also wary about the fact that the checkpoint is in Solidity but we use the IPLD resolver to grab it, which means if was part of the EVM data structures then it's only possible to interpret it by the EVM actor, so it would have to be shoveled over and stored in the BlockStore as IPLD, not as parts of the EVM KVstore, to make it accessible to the IPLD resolver. This would be trivial if we used built-in actors, but with EVM actors it's not.